WIP: containers: Cache docker image and node_modules in Semaphore unit test runs

martinpitt opened the pull request:
Use docker save/load to cache the test image in `$SEMAPHORE_CACHE_DIR`
if that is set. Ensure to not needlessly re-save an already cached
current image.

Also make the script work with `set -u`.

Change the hashbang to bash so that we can use the `${var/x/y}`
substitution, which is a bashism.

martinpitt commented:
The first test run should create the cache. Semaphore's current cache dir is completely empty right now, and interactive ssh sessions don't preserve the cache.

martinpitt commented:
Oh, I wonder why that wasn't working -- our semaphore config runs stuff through sudo, which cleans the environment. It's also not necessary, as `docker` magically works as user in semaphore (probably part of the docker group). I changed the config to drop sudo. @allisonkarlitskaya FYI

allisonkarlitskaya commented on containers/unit-tests/start:
```diff
@@ -22,6 +23,34 @@ if test -z "${docker:=$(which podman || which docker || true)}"; then
+fi
+
+set -x
+RC=0
```
I don't understand what's going on here.  This trick with conditionally setting RC would be useful if we were keeping `set -e`, but you get rid of it here.  In that case, why not just unconditionally capture RC=$? after?

martinpitt replied:
This does set -x, not set -e :-) I initially had it wrapped into set +e/set -e, but I find this a bit easier.

martinpitt commented on containers/unit-tests/start:
```diff
@@ -22,6 +23,34 @@ if test -z "${docker:=$(which podman || which docker || true)}"; then
+    cache_file="$SEMAPHORE_CACHE_DIR/docker/${image_esc}-$hash"
+    if ! [ -e "$cache_file" ]; then
+        # clean up older images
+        rm --verbose -f $SEMAPHORE_CACHE_DIR/docker/$image_esc-*
```
First> How do you define "other"? We e. g. don't want the :i386 test to remove the cached :latest image and vice versa. We could hardcode the set of known/current images, but I don't like that much either TBH.

Second> The trick is that we need to pull an image to get its hash, so I can't do it in the loading section. I can of course do the cleanup outside of the if and filter out the current one, but that's not going to make much practical difference. Usually the old one becomes out of date *exactly when* you have to write a new one.

atime: semaphore saves/restores the cache in between VMs; at first sight it seems it does not get restored correctly:
```
Semaphore (57 min left): ~/cockpit $ stat .semaphore-cache/docker/cockpit_unit-tests_latest-7a48e3e45d2f 
  File: '.semaphore-cache/docker/cockpit_unit-tests_latest-7a48e3e45d2f'
  Size: 1413226496	Blocks: 2760216    IO Block: 4096   regular file
Device: fc00h/64512d	Inode: 138001      Links: 1
Access: (0600/-rw-------)  Uid: ( 1003/  runner)   Gid: ( 1003/  runner)
Access: 2019-10-01 17:05:03.284000000 +0000
Modify: 2019-10-01 15:57:15.000000000 +0000
Change: 2019-10-01 17:05:14.663035000 +0000
 Birth: -
Semaphore (57 min left): ~/cockpit $ date
Tue Oct  1 17:08:14 UTC 2019
```

I. e. it has ctime and atime when the VM started, instead of the earlier run when it was created and used.

martinpitt commented:
The second run now did work; the log shows that it saved correctly, and I see the cache in an interactive ssh session. So I can take out the set -x debugging again. I'll move it to a separate script then (I slightly prefer one script, but I don't have a strong opinion.)

martinpitt commented:
Updated: I split out the caching (I kept it a bit generic to also accomodate the node_modules and possibly other stuff in the future), and made the cleanup more independent and also catch unused images (using mtime).

martinpitt commented:
The [current run](https://semaphoreci.com/cockpit-project/cockpit/branches/pull-request-12903/builds/5) shows that the cache is being used (in i386) and updated (in clang). However, something is still fishy -- an interactive ssh has an empty .semaphore-cache directory. I'll force-push again to trigger another build.

martinpitt commented on containers/unit-tests/semaphore-cache:
```diff
@@ -0,0 +1,39 @@
+
+        if [ -d "$SEMAPHORE_CACHE_DIR/docker" ]; then
+            echo "Cleaning old cached docker images..."
+            find "$SEMAPHORE_CACHE_DIR/docker" -type f -mtime +7 -print -delete
```
Wait, this is not sufficient now. This will only cover the "temporary tag now unused" case, but with the standard "update a tag" workflow this would unpack a recent and any older image for a week, the latter in vain. We *do* need to remove all other images immediately after saving, like in my original version. But we can keep that in addition for the former case.

allisonkarlitskaya replied:
I think I was a bit unclear on my proposal for cache invalidation.  My idea was to do the following two independent things, *always*:

 1) delete all hashes of our image name, except for our own hash
 2) delete absolutely everything that wasn't touched in over a week

martinpitt commented:
There's something odd going on.. in [run #6](https://semaphoreci.com/cockpit-project/cockpit/branches/pull-request-12903/builds/6), the gcc and i386 variants did not have a pre-existing cache, while the clang variant did. The [docs](https://semaphoreci.com/docs/caching-between-builds.html) don't mention this, but it feels like that the cache isn't shared between all semaphore runners, it might be local to a particular cloud region or similar? (Technically that would make a lot of sense)

I pushed another update to fix the cache cleanup as above.

martinpitt commented:
I updated this to also cache node_modules. Current version is a test commit that makes the semaphore test faster, for debugging. I want to validate the caching only.

martinpitt commented:
OK, [first run](https://semaphoreci.com/cockpit-project/cockpit/branches/pull-request-12903/builds/8) cached the dir. Force-pushing to do a second one which now ought to restore and use it.

martinpitt commented:
[Build #9](https://semaphoreci.com/cockpit-project/cockpit/branches/pull-request-12903/builds/9) used the cache for i386, but it didn't have a cache for the gcc and clang runs. Same non-deterministic "cache dir is not reliably preserved" issue as above..

martinpitt commented:
Closing for now, it seems this is too unreliable right now. My hunch is that semaphore messes up its cache dir with parallel jobs. I think it's a better direction to move to Semaphore 2.0 with its YAML based job descriptions, which also have container jobs built in (and thus it's their job to do the container image caching). 2.0 also has a much more fine-controlled caching, a tool similar to the one I implemented here. So I call this a dead end.
